Kubeadm: Add check certificate expiration command

fabriziopandini opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds a new command `kubeadm alpha certs check-expiration`

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/1563

**Does this PR introduce a user-facing change?**:
```release-note
Kubeadm: a new command `kubeadm alpha certs check-expiration` was created in order to help users in managing expiration for local PKI certificates
```

fabriziopandini commented:
/sig cluster-lifecycle
/area kubeadm
/priority important-soon
@kubernetes/sig-cluster-lifecycle-pr-reviews

/assign @rosti
/assign @timothysc

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -222,6 +227,33 @@ func printAvailableUpgrades(upgrades []upgrade.Upgrade, w io.Writer, isExternalE
+		fmt.Fprintf(w, "Following certificates managed by kubeadm are in the renewal windows (the %d days before the expiration date):\n\n", renewal.RenewalPeriodDays)
+		l.Write(w)
+		fmt.Fprintln(w, "\nkubeadm upgrade apply command is going to renew automatically certificates; if you want to skip certificate renewal")
+		fmt.Fprintln(w, "you can use pass \"--certificate-renewal=false\" flag to kubeadm upgrade apply plan.")
```
`kubeadm upgrade apply plan` -> `kubeadm upgrade apply`?

neolit123 commented on cmd/kubeadm/app/phases/certs/renewal/expiration.go:
```diff
@@ -0,0 +1,124 @@
+	"time"
+)
+
+// RenewalPeriodDays defines the leng of a period before the certificate expiration when renewal should be
```
`length`

neolit123 commented on cmd/kubeadm/app/phases/certs/renewal/expiration.go:
```diff
@@ -0,0 +1,124 @@
+// RenewalPeriodDays defines the leng of a period before the certificate expiration when renewal should be
+// planned and executed. This period consider that the default duration for kubeadm certificates is 365d,
+// the fact that kubeadm handles automatic certificate renewal during upgrades, and the release frequence of K8s.
+const RenewalPeriodDays = 180
```
ideally the constant for total time (365d) and this constant should be in the same file.

but i have a bit of concern here, as we are imposing a recommendation that not all users might agree with - 180d.

i think that if the user call `check-expiration` we should always print all certs that we manage and next to them in how much time they will expire. depending on a threshold of say 1 month we can tag some of the certs as recommended for renewal.

for `upgrade plan` we can only print the certs that are are about to expire in 1 month remaining.

this would be the UX i would like to see from kubeadm if i manage my certs with it.
WDYT @fabriziopandini

fabriziopandini replied:
>1 Month

I have no strong opinion and the threshold is a constant and it can be changed easily

However, based on issues and slack request of help, IMO 1 month is too short, especially considering users that are going to rely on kubeadm upgrade for automatic certs renewal.

In that case, we should consider the K8s release cycle of 90d, and according to the kubeadm survey, the fact that not all the user are upgrading frequently (hence the initial proposal of 180d).

Said that I will change to whatever value there is agreement on.

> I think that if the user calls check-expiration we should always print all certs that we manage and next to them in how much time they will expire. depending on a threshold of say _X_ month we can tag some of the certs as recommended for renewal.

> for upgrade plan we can only print the certs that are are about to expire in _X_ month remaining.

This is exactly how it works; see example above from the check-expiration command; upgrade output is the same but it contains only certificates set for expiration

neolit123 replied:
> Said that I will change to whatever value there is agreement on.

ok, let's discuss tomorrow (today):

fabriziopandini replied:
In the meantime 

> ideally the constant for total time (365d) and this constant should be in the same file.

was fixed

mauilion replied:
I agree that 180 days is pretty far out. 
Remember that the old rotated out cert is still valid. 
so the old cert is a liability for 180 days if we go this route. 30 or 60 days would be better.

fabriziopandini replied:
@mauilion this was discussed again in kubeadm office hours, and it was decided to drop the idea and always renew during upgrades

neolit123 reviewed (commented):
thanks for this work @fabriziopandini
this would be a very useful feature.

during the office hours we expressed that we might not want to add the extra info for `upgrade plan`, but i think it's fine and it's definitely nice to have...
.but we cannot add guarantees for the output format of this, as we don't have machine readable output there yet.

i've added a couple of extra comments.

rosti commented on cmd/kubeadm/app/phases/certs/renewal/expiration.go:
```diff
@@ -0,0 +1,124 @@
+	return e.ExpirationDate.Sub(time.Now())
+}
+
+func humanizeDuration(duration time.Duration) string {
```
Perhaps use [HumanDuration](https://github.com/kubernetes/kubernetes/blob/746404f82a28e55e0b76ffa7e40306fb88eb3317/staging/src/k8s.io/apimachinery/pkg/util/duration/duration.go#L48)?
kubeadm already uses `ShortHumanDuration` from the same package, so nothing new will be included.

fabriziopandini replied:
Good tip! Done

rosti commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -97,14 +99,17 @@ func runPlan(flags *planFlags, userVersion string) error {
 	}
 
+	// Create a certificate renewal manager
+	certsRenewMgr, _ := renewal.NewManager(&cfg.ClusterConfiguration, kubeadmconstants.KubernetesDir)
```
Unchecked error?

rosti reviewed (commented):
Thanks @fabriziopandini !
Looks OK on first pass.

fabriziopandini commented:
@rosti @neolit123 
Added/Fixed all the tests and addressed all the comments. Main points still open are:
a) 180d (current) vs 1m vs w whatever value we choose as renewalPeriod
b) show expiration in upgrade plan (current) or not

neolit123 commented:
> a) 180d (current) vs 1m vs w whatever value we choose as renewalPeriod

let's get feedback on that from the office hours today.

> b) show expiration in upgrade plan (current) or not

it's ok to print on plan.

stealthybox commented:
Feedback from office hours is to always renew certs on upgrade.
We'll should publish an `Action Required` release-note to warn kubeadm users who are copying these certs around that kubeadm is now automatically changing them on upgrade.

Also, we might consider an option to opt-out.

fabriziopandini commented:
@stealthybox @rosti @neolit123 
PR is now updated as discussed in the kubeadm office hours; an example of the command output is: 
```
root@kind-control-plane:/# kubeadm alpha certs check-expiration
CERTIFICATE                EXPIRES                  RESIDUAL TIME   EXTERNALLY MANAGED
admin.conf                 May 15, 2020 13:03 UTC   364d            true
apiserver                  May 15, 2020 13:00 UTC   364d            true
apiserver-etcd-client      May 15, 2020 13:00 UTC   364d            false
apiserver-kubelet-client   May 15, 2020 13:00 UTC   364d            true
controller-manager.conf    May 15, 2020 13:03 UTC   364d            true
etcd-healthcheck-client    May 15, 2020 13:00 UTC   364d            false
etcd-peer                  May 15, 2020 13:00 UTC   364d            false
etcd-server                May 15, 2020 13:00 UTC   364d            false
front-proxy-client         May 15, 2020 13:00 UTC   364d            false
scheduler.conf             May 15, 2020 13:03 UTC   364d            true
```
Since we are always renewing certificates during an upgrade, I dropped the changes to `kubeadm upgrade plan` too, thus simplifying the PR.

> We'll should publish an Action Required release-note to warn kubeadm users who are copying these certs around that kubeadm is now automatically changing them on upgrade.

The place for this "Action Required" is https://github.com/kubernetes/kubernetes/pull/76862; I will add this there

> Also, we might consider an option to opt-out.

Already done in https://github.com/kubernetes/kubernetes/pull/76862

fabriziopandini commented:
/test pull-kubernetes-e2e-gce

rosti reviewed (commented):
Thanks @fabriziopandini !
Looks great, only a tiny nit and I think we can merge it.

fabriziopandini committed:
certs-expiration

fabriziopandini committed:
autogenerated

fabriziopandini commented:
@rosti comments addressed!

dixudx reviewed (approved):
/lgtm

Thanks @fabianofranz
